crates/wdl-engine/src/eval/v1/workflow.rs

```diff
@@ -126,6 +132,197 @@ fn format_id(namespace: Option<&str>, target: &str, alias: &str, scatter_index:
+            expect_ty_match!(expr, Struct(s) => "Struct");
+            let members: Option<indexmap::IndexMap<String, Value>> = s
+                .items()
+                .map(|item| {
+                    let (name, val_expr) = item.name_value();
+                    let name_str = name.text().to_string();
+                    let member_type = struct_ty
+                        .members()
+                        .get(&name_str)
+                        .expect("member should exist in struct type");
+                    let val = extract_literal_value(member_type, &val_expr)?;
+                    Some((name_str, val))
+                })
+                .collect();
+            Some(Value::Compound(crate::CompoundValue::Object(
+                crate::Object::new(members?),
+            )))
+        }
+        Type::Object | Type::OptionalObject => {
+            expect_ty_match!(expr, Object(obj) => "Object");
+            let members: Option<indexmap::IndexMap<String, Value>> = obj
+                .items()
+                .map(|item| {
+                    let (name, val_expr) = item.name_value();
+                    let name_str = name.text().to_string();
+
+                    // Infer the type from the literal expression and recursively extract value
+                    let inferred_ty = infer_type_from_literal(&val_expr)?;
+                    let val = extract_literal_value(&inferred_ty, &val_expr)?;
+                    Some((name_str, val))
+                })
+                .collect();
+            Some(Value::Compound(crate::CompoundValue::Object(
+                crate::Object::new(members?),
+            )))
+        }
+        _ => None,
+    }
+}
+
+/// Builds enums from the document's analyzed enums.
+pub(crate) fn build_enums(document: &Document) -> HashMap<String, Enum> {
+    let mut enums = HashMap::new();
+
+    for (name, r#enum) in document.enums() {
+        let ty = r#enum.ty().expect("enum should have type");
+        let Type::Compound(CompoundType::Enum(enum_type), _) = ty else {
+            unreachable!("non-enum was returned by `document.enums()`");
+        };
+
+        let definition = r#enum.definition();
+        let variants: HashMap<String, Value> = definition
+            .variants()
+            .flat_map(|variant| {
+                let variant_name = variant.name().text().to_string();
+
+                // Get the variant's type from the analyzed enum type
+                let variant_type = enum_type
+                    .variants()
+                    .get(&variant_name)
+                    .expect("variant should exist in type");
+
+                // NOTE: `Type::Union` is used as a sentinel value in
+                // `wdl-analysis` to indicate that the variant type could not be
+                // parsed. As such, skip these variants in building values.
+                if matches!(variant_type, Type::Union) {
+                    return None;
+                }
```

Comment:
Using the variant type from the static analysis of the variant's expression is unnecessary and this is the only place that we actually get the value in the variant map (hence why I think it can just be an `IndexSet` of names from my much earlier comment).

We should be evaluating the constant expression and then coercing it to the enum variant's value type to produce the enum variant's value.